File: 149-MaxPointsonaLine/149-MaxPointsonaLine.py
# Last updated: 9/27/2025, 7:39:22 PM
from math import gcd
from collections import defaultdict

class Solution:
    def maxPoints(self, points: list[list[int]]) -> int:
        n = len(points)
        if n <= 2:
            return n

        result = 0

        for i in range(n):
            slopes = defaultdict(int)
            same_point = 1  
            curr_max = 0

            for j in range(i + 1, n):
                dx = points[j][0] - points[i][0]
                dy = points[j][1] - points[i][1]

            
                g = gcd(dx, dy)
                if g != 0:
                    dx //= g
                    dy //= g

          
                if dx == 0:
                    dy = 1
                elif dy == 0:
                    dx = 1
                elif dx < 0:  
                    dx = -dx
                    dy = -dy

                slopes[(dx, dy)] += 1
                curr_max = max(curr_max, slopes[(dx, dy)])

            result = max(result, curr_max + same_point)

        return result

File: 149-MaxPointsonaLine/149-MaxPointsonaLine.cpp
// Last updated: 9/27/2025, 7:37:22 PM
class Solution {
public:
    int maxPoints(vector<vector<int>>& points) {
        int n = points.size();
        if (n <= 2) return n;
        int result = 2;

        for (int i = 0; i < n; i++) {
            map<pair<int,int>, int> slopeCount;
            int localMax = 0;

            for (int j = i+1; j < n; j++) {
                int dx = points[j][0] - points[i][0];
                int dy = points[j][1] - points[i][1];

                int g = gcd(dx, dy);
                dx /= g;
                dy /= g;

                slopeCount[{dy, dx}]++;
                localMax = max(localMax, slopeCount[{dy, dx}]);
            }

            result = max(result, localMax + 1); // +1 for the base point itself
        }

        return result;
    }

private:
    int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }
};
